numba/array_analysis.py

```diff
@@ -0,0 +1,718 @@
+        self.typemap[size_var.name] = types.int64
+        getitem_node = ir.Expr.static_getitem(attr_var, i, const_var, var.loc)
+        self.calltypes[getitem_node] = None
+        getitem_assign = ir.Assign(getitem_node, size_var, var.loc)
+        out.append(getitem_assign)
+        return out
+
+    # lhs is array so rhs has to return array
+    def _analyze_rhs_classes(self, node):
+        if isinstance(node, ir.Arg):
+            assert self._isarray(node.name)
+            return self._add_array_corr(node.name)
+        elif isinstance(node, ir.Var):
+            return copy.copy(self.array_shape_classes[node.name])
+        elif isinstance(node, (ir.Global,ir.FreeVar)):
+            # XXX: currently, global variables are frozen in Numba (can change)
+            if isinstance(node.value, np.ndarray):
+                shape = node.value.shape
+                out_eqs = []
+                for c in shape:
+                    new_class = self._get_next_class_with_size(c)
+                    out_eqs.append(new_class)
+                return out_eqs
+        elif isinstance(node, ir.Expr):
+            if node.op=='unary' and node.fn in UNARY_MAP_OP:
+                assert isinstance(node.value, ir.Var)
+                in_var = node.value.name
+                assert self._isarray(in_var)
+                return copy.copy(self.array_shape_classes[in_var])
+            elif node.op=='binop' and node.fn in BINARY_MAP_OP:
+                arg1 = node.lhs.name
+                arg2 = node.rhs.name
+                return self._broadcast_and_match_shapes([arg1, arg2])
+            elif node.op=='inplace_binop' and node.immutable_fn in BINARY_MAP_OP:
+                arg1 = node.lhs.name
+                arg2 = node.rhs.name
+                return self._broadcast_and_match_shapes([arg1, arg2])
+            elif node.op=='arrayexpr':
+                # set to remove duplicates
+                args = {v.name for v in node.list_vars()}
+                return self._broadcast_and_match_shapes(list(args))
+            elif node.op=='cast':
+                return copy.copy(self.array_shape_classes[node.value.name])
+            elif node.op=='call':
+                call_name = 'NULL'
+                args = copy.copy(node.args)
+                if node.func.name in self.map_calls:
+                    return copy.copy(self.array_shape_classes[args[0].name])
+                if node.func.name in self.numpy_calls.keys():
+                    call_name = self.numpy_calls[node.func.name]
+                elif node.func.name in self.array_attr_calls.keys():
+                    call_name, arr = self.array_attr_calls[node.func.name]
+                    args.insert(0,arr)
+                if call_name is not 'NULL':
+                    return self._analyze_np_call(call_name, args, dict(node.kws))
+                else:
+                    if config.DEBUG_ARRAY_OPT==1:
+                        print("can't find shape for unknown call:", node)
```

Comment:
No need to raise since this is not a failure and analysis can continue. It might just limit optimizations later. I added comments to explain.